akka-stream/src/main/scala/akka/stream/remote/serialization/StreamRefSerializer.scala

```diff
@@ -0,0 +1,145 @@
+    case _: StreamRefs.SequencedOnNext[_] ⇒ SequencedOnNextManifest
+    case _: StreamRefs.CumulativeDemand   ⇒ CumulativeDemandManifest
+    // refs
+    case _: SourceRef[_]                  ⇒ SourceRefManifest
+    case _: SinkRef[_]                    ⇒ SinkRefManifest
+  }
+
+  override def toBinary(o: AnyRef): Array[Byte] = o match {
+    // protocol
+    case o: StreamRefs.SequencedOnNext[_] ⇒ serializeSequencedOnNext(o).toByteArray
+    case d: StreamRefs.CumulativeDemand   ⇒ serializeCumulativeDemand(d).toByteArray
+    // refs
+    case ref: SourceRef[_]                ⇒ serializeSourceRef(ref).toByteArray
+    case ref: SinkRef[_]                  ⇒ serializeSinkRef(ref).toByteArray
+  }
+
+  override def fromBinary(bytes: Array[Byte], manifest: String): AnyRef = manifest match {
+    // protocol
+    case SequencedOnNextManifest  ⇒ deserializeSequencedOnNext(bytes)
+    case CumulativeDemandManifest ⇒ deserializeCumulativeDemand(bytes)
+    // refs
+    case SinkRefManifest          ⇒ deserializeSinkRef(bytes)
+    case SourceRefManifest        ⇒ ???
+  }
+
+  // -----
+
+  private def serializeCumulativeDemand(d: StreamRefs.CumulativeDemand): StreamRefContainers.CumulativeDemand = {
+    StreamRefContainers.CumulativeDemand.newBuilder()
+      .setSeqNr(d.seqNr)
+      .build()
+  }
+
+  private def serializeSequencedOnNext(o: StreamRefs.SequencedOnNext[_]) = {
+    val p = o.payload.asInstanceOf[AnyRef]
+    val msgSerializer = serialization.findSerializerFor(p)
+
+    val payloadBuilder = StreamRefContainers.Payload.newBuilder()
+      .setEnclosedMessage(ByteString.copyFrom(msgSerializer.toBinary(p)))
+      .setSerializerId(msgSerializer.identifier)
+
+    msgSerializer match {
+      case ser2: SerializerWithStringManifest ⇒
+        val manifest = ser2.manifest(p)
+        if (manifest != "")
+          payloadBuilder.setMessageManifest(ByteString.copyFromUtf8(manifest))
+      case _ ⇒
+        if (msgSerializer.includeManifest)
+          payloadBuilder.setMessageManifest(ByteString.copyFromUtf8(p.getClass.getName))
+    }
+
+    StreamRefContainers.SequencedOnNext.newBuilder()
+      .setSeqNr(o.seqNr)
+      .setPayload(payloadBuilder.build())
+      .build()
+  }
+
+  private def serializeSinkRef(sink: SinkRef[_]): StreamRefContainers.ActorRef = {
+    val make = () ⇒ StreamRefContainers.ActorRef.newBuilder()
+      .setPath(Serialization.serializedActorPath(sink.targetRef))
+      .build()
+
+    // Serialize actor references with full address information (defaultAddress).
+    // When sending remote messages currentTransportInformation is already set,
+    // but when serializing for digests it must be set here.
+    val it = if (Serialization.currentTransportInformation.value == null)
+      Serialization.currentTransportInformation.withValue(transportInformation) {
+        println(s"INSIDE Serialization.serializedActorPath(sink.targetRef) = ${Serialization.serializedActorPath(sink.targetRef)}")
+        make()
```

Comment:
Is this right? It seems I'm setting the wrong address here for some reason...

```
LOCAL  ::::: akka.tcp://StreamRefsSpec@localhost:54510
REMOTE ::::: akka.tcp://RemoteSystem@localhost:54513

=== serializeSinkRef it = path: "akka.tcp://StreamRefsSpec@localhost:54510/user/actor/StreamSupervisor-1/$$a#43440242"
```

but the SinkRef was created in the remote system... Hm :\